File: source/utils/random/independantGaussianRNG.cpp

/*__________________________________________________
 * utils/random/
 * independantGaussianRNG.cpp
 *__________________________________________________
 * author        : colonel
 * last modified : 2016/9/16
 *__________________________________________________
 *
 * implementation of class IndependantGaussianRNG
 *
 */

#ifndef UTILS_RANDOM_INDEPENDANTGAUSSIANRNG_CPP
#define UTILS_RANDOM_INDEPENDANTGAUSSIANRNG_CPP

#include "independantGaussianRNG.h"

namespace utils
{

    namespace random
    {

        // set common random number generator
        template <std::size_t N, typename Real, class RNG>
            void IndependantGaussianRNG<N, Real, RNG> ::
            setRNG(RNG * t_rng)
            {
                m_rng = t_rng ;
            }

        // set distribution parameters
        template <std::size_t N, typename Real, class RNG>
            void IndependantGaussianRNG<N, Real, RNG> ::
            setParameters(const Real * t_mean, const Real * t_var)
            {
                for(std::size_t i = 0 ; i < N ; ++i)
                {
                    m_distribution[i] = Normal <Real>(t_mean[i], t_var[i]) ;
                }
            }

        // add error to t_vector
        template <std::size_t N, typename Real, class RNG>
            void IndependantGaussianRNG<N, Real, RNG> ::
            addError(Real * t_vector)
            {
                for(std::size_t i = 0 ; i < N ; ++i)
                {
                    t_vector[i] += m_distribution[i](*m_rng) ;
                }
            }

    }

}

#endif // UTILS_RANDOM_INDEPENDANTGAUSSIANRNG_CPP

File: python/pyLorenz/utils/random/independantGaussianRNG.py
#! /usr/bin/env python

#__________________________________________________
# pyLorenz/utils/random/
# independantGaussianRNG.py
#__________________________________________________
# author        : colonel
# last modified : 2016/9/20
#__________________________________________________
#
# class to handle a gaussian random number generator
# when the covariance matrix is diagonal
#

import numpy as np
import numpy.random as rnd

#__________________________________________________

class IndependantGaussianRNG:

    #_________________________

    def __init__(self):
        self.m_relSigmaMin = 1.0e-5
        self.setParameters()

    #_________________________

    def setParameters(self, t_mean = np.zeros(0), t_sigma = np.zeros(0)):
        # set parameter for the gaussian distribution
        self.m_mean  = t_mean
        self.m_sigma = np.maximum(t_sigma, self.m_relSigmaMin)
        self.m_exact = self.m_sigma < self.m_relSigmaMin

    #_________________________

    def addError(self, t_x):
        # add random error to x according to the gaussian distribution
        shape = t_x.shape
        if len(shape) == 1:
            # for one vector
            t_x += self.drawSample()
        else:
            # for multiple vectors
            t_x += self.drawSamples(shape[0])

    #_________________________

    def drawSample(self):
        # draw one sample according to the gaussian distribution
        return rnd.normal(self.m_mean, self.m_sigma) * ( 1.0 - self.m_exact ) + self.m_mean * self.m_exact

    #_________________________

    def drawSamples(self, t_Ns):
        # draw t_Ns samples according to the gaussian distribution
        return ( rnd.normal( np.tile(self.m_mean, (t_Ns, 1)) , np.tile(self.m_sigma, (t_Ns, 1)) ) * ( 1.0 - np.tile(self.m_exact, (t_Ns, 1)) ) +
                np.tile(self.m_mean*self.m_exact, (t_Ns, 1)) )

#__________________________________________________

